x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleVersionPlugin.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.compat;
+
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.plugins.RestCompatibilityPlugin;
+import org.elasticsearch.rest.RestStatus;
+
+public class CompatibleVersionPlugin extends Plugin implements RestCompatibilityPlugin {
+
+    @Override
+    public Version getCompatibleVersion(@Nullable String acceptHeader, @Nullable String contentTypeHeader, boolean hasContent) {
+        Byte aVersion = XContentType.parseVersion(acceptHeader);
+        byte acceptVersion = aVersion == null ? Version.CURRENT.major : Integer.valueOf(aVersion).byteValue();
+        Byte cVersion = XContentType.parseVersion(contentTypeHeader);
+        byte contentTypeVersion = cVersion == null ? Version.CURRENT.major : Integer.valueOf(cVersion).byteValue();
```

Comment:
this only supports versioned media types defined in `XContentType` so json, yaml, cbor, smile
what if however someone sends a request like
```
curl --request POST \
  --url http://localhost:9200/_sql \
  --header 'accept: application/vnd.elasticsearch+csv;compatible-with=7' \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/vnd.elasticsearch+json;compatible-with=7' \
  --header 'x-opaque-id: 123' \
  --data '{
  "query": "SELECT * FROM library WHERE release_date < \u00272000-01-01\u0027"
}
'
```
the current logic is not able to parse `accept: application/vnd.elasticsearch+csv;compatible-with=7` so sets `acceptVersion = 8` however it can parse `content-type: application/vnd.elasticsearch+json;compatible-with=7'` so `contentTypeVersion = 7` that leads to an exception https://github.com/elastic/elasticsearch/pull/60516/files#diff-793fe1b13ffa171bfc74dab865948b08R47

How will clients send these headers? will client be aware to which API should they send versioned media types?
@elastic/es-clients any thoughts?

To make it work now the request would have to be sent without versions.
```
curl --request POST \
  --url http://localhost:9200/_sql \
  --header 'accept: application/csv' \
  --header 'content-type: application/json' \
  --header 'x-opaque-id: 123' \
  --data '{
  "query": "SELECT * FROM library WHERE release_date < \u00272000-01-01\u0027"
}
'
```

Reply:
The clients are running with the assumption that we can send the `vendor mimetype` to all API's.

That way a client running `7.11` can start sending `application/vnd.elasticsearch+json;compatible-with=7` to API's even if that API does not have any special backward compatibility handling yet.

If the user upgrades the server first to `7.11` and the API now does have backward compatibility baked in nothing really changes when the client is updated afterwards. 


If the user updates to `8.x` before the API gets `bwc` handling the client will still send and receive v7 mime types. 

```
-> accept: application/vnd.elasticsearch+json;compatible-with=7
-> content-type: application/vnd.elasticsearch+json;compatible-with=7
``` 

If the client upgrades to 8.x:

```
-> accept: application/vnd.elasticsearch+json;compatible-with=8
->  content-type: application/vnd.elasticsearch+json;compatible-with=8
``` 

If nothing changes in the server `10.x` and the API **still** has no backward compatibly layer. (and the client is still on 8.x)

```
-> accept: application/vnd.elasticsearch+json;compatible-with=8
-> content-type: application/vnd.elasticsearch+json;compatible-with=8
``` 

Will fail, even if nothing actually changed in between these different versions. Adhering to supported backward compatibility guides.

> the current logic is not able to parse accept: application/vnd.elasticsearch+csv;compatible-with=7 so sets acceptVersion = 8 however it can parse content-type: application/vnd.elasticsearch+json;compatible-with=7' so contentTypeVersion = 7 

What are the cases where it can happen that `json` will have a compatible response but `csv` does not? Don't they flow through the same xcontent builders/parsers? 

The behavior of setting `acceptVersion` to something different then what the user/client explicitly provided should not happen. The resulting version mismatch error will be very confusing.